coreax/metrics.py

```diff
@@ -129,3 +132,141 @@ def compute(
+
+        where :math:`\hat{\mu}^{(1)},\hat{\mu}^{(2)}` are the conditional mean
+        embeddings estimated with :math:`\mathcal{D}^{(1)}` and
+        :math:`\mathcal{D}^{(2)}` respectively, and :math:`\mathcal{H}_k,\mathcal{H}_l`
+        are the RKHSs corresponding to the kernel functions
+        :math:`k: \mathbb{R}^d \times \mathbb{R}^d \rightarrow \mathbb{R}` and
+        :math:`l: \mathbb{R}^p \times \mathbb{R}^p \rightarrow \mathbb{R}` respectively.
+
+        :param reference_data: The original supervised dataset :math:`\mathcal{D}^{(1)}=
+            \{(x_i, y_i)\}_{i=1}^n` of ``n`` pairs with :math:`x\in\mathbb{R}^d` and
+            :math:`y\in\mathbb{R}^p`
+        :param comparison_data: Supervised dataset
+            :math:`\mathcal{D}^{(2)} = \{(\tilde{x}_i, \tilde{y}_i)\}_{i=1}^n` of ``m``
+            pairs with :math:`\tilde{x}\in\mathbb{R}^d` and
+            :math:`\tilde{y}\in\mathbb{R}^p`
+        :return: Conditional maximum mean discrepancy as a 0-dimensional array
+        """
+        del kwargs
+        # Extract features and responses from reference and comparison datasets
+        x1 = reference_data.data
+        y1 = reference_data.supervision
+        x2 = comparison_data.data
+        y2 = comparison_data.supervision
+
+        # Compute feature kernel gramians
+        feature_gramian_1 = self.feature_kernel.compute(x1, x1)
+        feature_gramian_2 = self.feature_kernel.compute(x2, x2)
+
+        # Invert feature kernel gramians
+        if self.inverse_approximator is None:
+            inverse_approximator = LeastSquareApproximator(jr.key(2_024))
+        else:
+            inverse_approximator = self.inverse_approximator
+        inverse_feature_gramian_1 = inverse_approximator.approximate(
+            array=feature_gramian_1,
+            regularisation_parameter=self.regularisation_parameter,
+            identity=jnp.eye(feature_gramian_1.shape[0]),
+        )
+
+        inverse_feature_gramian_2 = inverse_approximator.approximate(
+            array=feature_gramian_2,
+            regularisation_parameter=self.regularisation_parameter,
+            identity=jnp.eye(feature_gramian_2.shape[0]),
+        )
+
+        # Compute each term in the CMMD
+        term_1 = (
+            inverse_feature_gramian_1
+            @ self.response_kernel.compute(y1, y1)
+            @ inverse_feature_gramian_1
+            @ feature_gramian_1
+        )
+        term_2 = (
+            inverse_feature_gramian_2
+            @ self.response_kernel.compute(y2, y2)
+            @ inverse_feature_gramian_2
+            @ feature_gramian_2
+        )
+        term_3 = (
+            inverse_feature_gramian_1
+            @ self.response_kernel.compute(y1, y2)
+            @ inverse_feature_gramian_2
+            @ self.feature_kernel.compute(x2, x1)
+        )
+
+        # Compute CMMD
+        squared_cmmd = jnp.trace(term_1) + jnp.trace(term_2) - 2 * jnp.trace(term_3)
```

Comment:
Given symmetric matrices $A, B \in \mathbb{R}^{n\times n}$, we have $\text{Tr}(AB) := \sum_{i=1}^n(AB)_{ii} = \sum_{i,j=1}^n a_{ij}{bji} = \sum_{i,j=1}^n  (A \cdot B)_{ij}$ i.e. the sum of their Hadamard product. So we could compute the trace this way with no wasted work. But I remember testing this previously and not seeing any time improvement. Will double check and report back

Reply:
Top is current way, bottom is Hadamard way 
![image](https://github.com/gchq/coreax/assets/167978891/f398b35c-46fe-4118-99bc-62698f566569)
Does seem like the Hadamard way saves time. Not sure what happened last time!

Added.

Reply:
This is a nice win!

Reply:
Interesting, can you post benchmark times for the complete CMMD metric computation (when JIT compiled) for both approaches, and with 32bit and 64bit default precision. I'll have a quick think why this might be.

Reply:
Elaborating on the above, when $A, B$ are rectangular matrices, we have
```math
\text{Tr}(A^TB) := \sum_{i=1}^{n} (A^TB)_{ii} = \sum_{i=1}^{n} \sum_{m=1}^{n}a_{ij}b_{ij} = \sum_{i,j=1}^{n}  (A^T \odot B)_{ij}
```

Reply:
We could just compute the third term as a trace but this doesn't remove the numerical instability issues that the tests are flagging i.e. computing the CMMD on a dataset with itself seems much more likely to be non-zero or negative with the Hadamard method based on some fiddling I've done.

Reply:
Will do! In the office for a team day today so not sure how much coding I will be able to get done 

> Interesting, can you post benchmark times for the complete CMMD metric computation (when JIT compiled) for both approaches, and with 32bit and 64bit default precision. I'll have a quick think why this might be.

Will do! In the office today for a team day so not sure how much coding I will be doing.

Reply:
If you group the matrix multiplies as so: `jnp.sum((A @ B).T * (C @ D))`, using the brackets to ensure the matmul is before the element wise multiply, do we see any appreciable changes in speed? Regarding the shapes, aren't the matrices always square?

Reply:
The third term involves rectangular matrices, and the order of operations which you suggested is what I had implemented. 

By stability I mean that computations that were working before i.e. CMMD(x, x) = 0 from the tests are now giving non-zero or negative answrs indicating we are having numerically instability issues where there weren'tpreviously.

Reply:
Ok, that all makes sense. If you want to get this PR merged ASAP, open a new ticket to revist this (potential) optimisation at a future date.

Reply:
Okay, CMMD is the original implementation, CMMD2 is the Hadamard implementation and CMMD3 is a mix of the two where we avoid the transpose in the third term by using the trace. These are the timings:

![image](https://github.com/gchq/coreax/assets/167978891/b46d3ac8-31f2-4a5d-9d0b-9da34b7fcd11)

Clearly the Hadamard method is the fastest, and the original is the slowest, though the difference isn't huge.

Running the tests on `float32` we can see the original CMMD is the most stable (in the sense I described above):

![image](https://github.com/gchq/coreax/assets/167978891/626e7b52-8554-4633-b563-d028ddeac232)

Enabling `float64` results in lots of negative values of CMMD for CMMD2 and CMMD3 (resulting in `nan`s), but none for the original implementation:

![image](https://github.com/gchq/coreax/assets/167978891/d2f6d785-adcc-4da2-8b6e-fe9356befaa2)

Ignore the approximator tests, that is likely a bug with enforcing 32 bit somewhere in the code.

Reply:
I think this supports sticking with the original implementation as the speed improvements are not order of magnitudes.

Reply:
I'm happy with that. If you still have the code, could you perform the benchmarks on a GPU as well. I have a feeling that the specialised matrix-multiply hardware (E.G. Nvidia's Tensor Cores) might allow the original implementation to be faster on these devices anyway.

Reply:
Sure, I'll set up an EC2 hopefully today and get back to you!